File: sdks/nodejs/src/frameworks/llamaindex.ts
import type { CyberArmorClient } from '../client';
import { isAllowed, PolicyViolationError } from '../policy/decision';

type LlamaIndexLike = {
  complete?: (prompt: string, ...rest: unknown[]) => Promise<unknown>;
  acomplete?: (prompt: string, ...rest: unknown[]) => Promise<unknown>;
};

export function protectLlamaIndexLLM<T extends LlamaIndexLike>(
  llm: T,
  ca: CyberArmorClient,
  model = 'unknown',
): T {
  const wrapped = llm as T;
  if (wrapped.complete) {
    const original = wrapped.complete.bind(wrapped);
    wrapped.complete = (async (prompt: string, ...rest: unknown[]) => {
      const decision = await ca.evaluatePolicy({
        action: 'llm_call',
        model,
        promptText: String(prompt ?? '').slice(0, 2000),
      });
      if (!isAllowed(decision) && ca.config.enforceMode === 'block') {
        throw new PolicyViolationError(decision);
      }
      const start = Date.now();
      const out = await original(prompt, ...rest);
      ca.emitEvent('llm_call', {
        framework: 'llamaindex',
        model,
        outcome: 'success',
        latencyMs: Date.now() - start,
      });
      return out;
    }) as T['complete'];
  }
  if (wrapped.acomplete) {
    const original = wrapped.acomplete.bind(wrapped);
    wrapped.acomplete = (async (prompt: string, ...rest: unknown[]) => {
      const decision = await ca.evaluatePolicy({
        action: 'llm_call',
        model,
        promptText: String(prompt ?? '').slice(0, 2000),
      });
      if (!isAllowed(decision) && ca.config.enforceMode === 'block') {
        throw new PolicyViolationError(decision);
      }
      return original(prompt, ...rest);
    }) as T['acomplete'];
  }
  return wrapped;
}

File: sdks/go/frameworks/llamaindex/llamaindex.go
package llamaindex

import (
	"context"
	"fmt"
	"time"

	cyberarmor "github.com/cyberarmor-ai/cyberarmor-go"
)

// QueryInvoker represents a framework-native query function (LlamaIndex-style).
type QueryInvoker func(ctx context.Context, query string) (string, error)

// Guard wraps a query invoker with CyberArmor policy enforcement + audit emission.
func Guard(
	client *cyberarmor.Client,
	provider string,
	model string,
	tenantID string,
	invoker QueryInvoker,
) QueryInvoker {
	return func(ctx context.Context, query string) (string, error) {
		start := time.Now()

		decision, err := client.EvaluatePolicy(ctx, cyberarmor.EvaluatePolicyOptions{
			Action:     "framework.llamaindex.query",
			Provider:   provider,
			Model:      model,
			PromptText: query,
		})
		if err != nil && !client.IsFailOpen() {
			return "", fmt.Errorf("policy evaluation failed: %w", err)
		}
		if err == nil && !decision.IsAllowed() && client.GetEnforceMode() == "block" {
			client.EmitEvent("framework.llamaindex.blocked", map[string]any{
				"provider": provider,
				"model":    model,
				"tenant_id": tenantID,
				"decision": decision.Type,
			})
			return "", &cyberarmor.PolicyViolationError{Decision: decision}
		}

		out, invokeErr := invoker(ctx, query)
		client.EmitEvent("framework.llamaindex.query", map[string]any{
			"provider":   provider,
			"model":      model,
			"tenant_id":  tenantID,
			"latency_ms": int(time.Since(start).Milliseconds()),
			"outcome":    map[bool]string{true: "error", false: "success"}[invokeErr != nil],
		})
		return out, invokeErr
	}
}
